From 1e185bb0c2097e897c8b8020cc19f8799755703d Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 19 Feb 2025 15:45:27 +0100
Subject: [PATCH 1592/1695] plat-stm32mp*: move threaded interrupt from PWR to
 pmic client

The handling of the threaded interrupt should belong to the code
of the client.

Remove it from PWR driver and move it to the PMIC driver.

While there:
- drop the now unused macro PWR_WKUP_FLAG_THREADED;
- tag the PWR interrupt with DECLARE_KEEP_PAGER.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1dba52950b1422459840d004efef8246fd867f2e plat-stm32mp*: move threaded interrupt from PWR to pmic client]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I09afa55b020fc4e61a3525d3a9ec355380492281
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/392424
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 .../arm/plat-stm32mp1/drivers/stm32mp1_pwr.h  |   2 -
 .../plat-stm32mp1/drivers/stm32mp1_pwr_irq.c  |  55 +---------
 .../plat-stm32mp1/drivers/stm32mp1_stpmic1.c  |  86 ++++++++++++++-
 .../arm/plat-stm32mp2/drivers/stm32mp25_pwr.h |   2 -
 .../plat-stm32mp2/drivers/stm32mp25_pwr_irq.c |  57 +---------
 core/drivers/stm32mp_pmic2.c                  | 102 +++++++++++++++++-
 6 files changed, 185 insertions(+), 119 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h
index 7b2c606bc..ad4aa33e7 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h
@@ -70,11 +70,9 @@
  * Flags for PWR wakeup event management
  * PWR_WKUP_FLAG_RISING - Detect event on signal rising edge
  * PWR_WKUP_FLAG_FALLING - Detect event on signal falling edge
- * PWR_WKUP_FLAG_THREADED - Notify event in the threaded context
  */
 #define PWR_WKUP_FLAG_RISING	0
 #define PWR_WKUP_FLAG_FALLING	BIT(0)
-#define PWR_WKUP_FLAG_THREADED	BIT(1)
 
 enum pwr_regulator {
 	PWR_REG11 = 0,
diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
index 49101dee1..bdc44b53c 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
@@ -51,8 +51,6 @@ struct stm32_pwr_data {
 	struct itr_handler *hdl[PWR_NB_WAKEUPPINS];
 	struct itr_handler *parent_hdl;
 	struct stm32_exti_pdata *exti;
-	bool threaded[PWR_NB_WAKEUPPINS];
-	bool pending[PWR_NB_WAKEUPPINS];
 	unsigned int spinlock;
 };
 
@@ -110,45 +108,6 @@ static enum itr_return pwr_it_call_handler(struct stm32_pwr_data *priv,
 	return ITRR_HANDLED;
 }
 
-static enum itr_return pwr_it_threaded_handler(void)
-{
-	struct stm32_pwr_data *priv = pwr_data;
-	uint32_t i = 0;
-
-	VERBOSE_PWR("");
-
-	for (i = 0; i < PWR_NB_WAKEUPPINS; i++) {
-		if (priv->pending[i]) {
-			VERBOSE_PWR("handle pending wkup irq:%"PRIu32, i);
-			priv->pending[i] = false;
-			pwr_it_call_handler(priv, i);
-		}
-	}
-
-	return ITRR_HANDLED;
-}
-
-static void yielding_stm32_pwr_notif(struct notif_driver *ndrv __unused,
-				     enum notif_event ev)
-{
-	switch (ev) {
-	case NOTIF_EVENT_DO_BOTTOM_HALF:
-		VERBOSE_PWR("PWR Bottom half");
-		pwr_it_threaded_handler();
-		break;
-	case NOTIF_EVENT_STOPPED:
-		VERBOSE_PWR("PWR notif stopped");
-		break;
-	default:
-		EMSG("Unknown event %d", ev);
-		panic();
-	}
-}
-
-struct notif_driver stm32_pwr_notif = {
-	.yielding_cb = yielding_stm32_pwr_notif,
-};
-
 static enum itr_return pwr_it_handler(struct itr_handler *handler)
 {
 	struct stm32_pwr_data *priv = (struct stm32_pwr_data *)handler->data;
@@ -168,12 +127,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 			/* Ack IRQ */
 			io_setbits32(priv->base + WKUPCR, BIT(i));
 
-			if (priv->threaded[i] && notif_async_is_started()) {
-				priv->pending[i] = true;
-				notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
-			} else {
-				pwr_it_call_handler(priv, i);
-			}
+			pwr_it_call_handler(priv, i);
 		}
 	}
 
@@ -181,6 +135,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 
 	return ITRR_HANDLED;
 }
+DECLARE_KEEP_PAGER(pwr_it_handler);
 
 static TEE_Result
 stm32_pwr_irq_set_pull_config(size_t it, enum wkup_pull_setting config)
@@ -530,9 +485,6 @@ static TEE_Result stm32mp1_pwr_itr_add(const void *fdt, int wp_node,
 	if (!itr_free)
 		return TEE_ERROR_GENERIC;
 
-	if (hdl->flags & PWR_WKUP_FLAG_THREADED)
-		priv->threaded[it] = true;
-
 	res = gpio_dt_get_by_index(fdt, wp_node, it, "wakeup", &gpio);
 	if (res) {
 		priv->hdl[it] = NULL;
@@ -633,9 +585,6 @@ stm32mp1_pwr_irq_probe(const void *fdt, int node,
 
 	interrupt_enable(priv->parent_hdl->chip, priv->parent_hdl->it);
 
-	if (IS_ENABLED(CFG_CORE_ASYNC_NOTIF))
-		notif_register_driver(&stm32_pwr_notif);
-
 	res = dt_driver_register_provider(fdt, node, NULL,
 					  NULL, DT_DRIVER_NOTYPE);
 	if (res)
diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
index 209470d43..63a108c12 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
@@ -77,6 +77,9 @@ static SLIST_HEAD(pmic_it_handle_head, pmic_it_handle_s) pmic_it_handle_list =
 /* CPU voltage supplier if found */
 static char cpu_supply_name[PMIC_REGU_SUPPLY_NAME_LEN];
 
+/* Flag set when bottom half thread is to enable PMIC interrupts */
+static bool start_stpmic1_interrupt;
+
 static bool pmic_is_secure(void)
 {
 	assert(pmic_status != -1);
@@ -755,7 +758,7 @@ static TEE_Result initialize_pmic(const void *fdt, int pmic_node)
 	return TEE_SUCCESS;
 }
 
-static enum itr_return stpmic1_irq_handler(struct itr_handler *handler __unused)
+static void pmic_thread_irq_handler(void)
 {
 	uint8_t read_val = 0U;
 	unsigned int i = 0U;
@@ -804,9 +807,82 @@ static enum itr_return stpmic1_irq_handler(struct itr_handler *handler __unused)
 		 (latch4 != 0));
 
 	stm32mp_put_pmic();
+}
+
+static void start_stop_stpmic1_interrupt(bool start_not_stop)
+{
+	struct pmic_it_handle_s *prv = NULL;
+	uint8_t val = 0;
+
+	stm32mp_get_pmic();
+
+	/* Enable requested interrupt */
+	SLIST_FOREACH(prv, &pmic_it_handle_list, link) {
+		if (stpmic1_register_read(prv->pmic_reg, &val))
+			panic();
+
+		if (start_not_stop)
+			val |= BIT(prv->pmic_bit);
+		else
+			val &= ~BIT(prv->pmic_bit);
+
+		if (stpmic1_register_write(prv->pmic_reg, val))
+			panic();
+	}
+
+	stm32mp_put_pmic();
+}
+
+static void yielding_stm32mp1_stpmic1_notif(struct notif_driver *ndrv __unused,
+					    enum notif_event ev)
+{
+	switch (ev) {
+	case NOTIF_EVENT_DO_BOTTOM_HALF:
+		if (start_stpmic1_interrupt) {
+			IMSG("STPMIC1 notif started");
+			start_stpmic1_interrupt = false;
+			start_stop_stpmic1_interrupt(true /*start_not_stop*/);
+		}
+
+		pmic_thread_irq_handler();
+		break;
+	case NOTIF_EVENT_STOPPED:
+		IMSG("STPMIC1 notif stopped");
+		start_stpmic1_interrupt = false;
+		start_stop_stpmic1_interrupt(false /*!start_not_stop*/);
+		break;
+	default:
+		EMSG("Unknown event %d", ev);
+		panic();
+	}
+}
+
+static bool atomic_stm32mp1_stpmic1_notif(struct notif_driver *ndrv __unused,
+					  enum notif_event ev __maybe_unused)
+{
+	assert(ev == NOTIF_EVENT_STARTED);
+
+	/* Ask for the bottom half to enable PMIC interrupts */
+	start_stpmic1_interrupt = true;
+
+	return true;
+}
+DECLARE_KEEP_PAGER(atomic_stm32mp1_stpmic1_notif);
+
+/* Structure below is not const to prevent it spreads in the unpaged sections */
+static struct notif_driver stm32mp1_stpmic1_notif = {
+	.atomic_cb = atomic_stm32mp1_stpmic1_notif,
+	.yielding_cb = yielding_stm32mp1_stpmic1_notif,
+};
+
+static enum itr_return pmic_it_handler(struct itr_handler *handler __unused)
+{
+	if (notif_async_is_started())
+		notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
 
 	return ITRR_HANDLED;
 }
+DECLARE_KEEP_PAGER(pmic_it_handler);
 
 static TEE_Result stm32_pmic_init_it(const void *fdt, int node)
 {
@@ -820,6 +896,9 @@ static TEE_Result stm32_pmic_init_it(const void *fdt, int node)
 	int wakeup_parent_node = 0;
 	int len = 0;
 
+	if (IS_ENABLED(CFG_CORE_ASYNC_NOTIF))
+		notif_register_driver(&stm32mp1_stpmic1_notif);
+
 	cuint = fdt_getprop(fdt, node, "wakeup-parent", &len);
 	if (!cuint || len != sizeof(uint32_t))
 		panic("Missing wakeup-parent");
@@ -893,9 +972,8 @@ static TEE_Result stm32_pmic_init_it(const void *fdt, int node)
 	}
 
 	res = stm32mp1_pwr_itr_alloc_add(fdt, wakeup_parent_node, pwr_it,
-					 stpmic1_irq_handler,
-					 PWR_WKUP_FLAG_FALLING |
-					 PWR_WKUP_FLAG_THREADED,
+					 pmic_it_handler,
+					 PWR_WKUP_FLAG_FALLING,
 					 NULL, &hdl);
 	if (res)
 		panic("pmic: Could not allocate itr");
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
index ebe19996d..2501e3960 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
@@ -16,11 +16,9 @@
  * Flags for PWR wakeup event management
  * PWR_WKUP_FLAG_RISING - Detect event on signal rising edge
  * PWR_WKUP_FLAG_FALLING - Detect event on signal falling edge
- * PWR_WKUP_FLAG_THREADED - Notify event in the threaded context
  */
 #define PWR_WKUP_FLAG_RISING	0
 #define PWR_WKUP_FLAG_FALLING	BIT(0)
-#define PWR_WKUP_FLAG_THREADED	BIT(1)
 
 /* wakeup-pins irq chip */
 enum pwr_wkup_pins {
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
index 2bb5591c6..caface1ce 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
@@ -53,8 +53,6 @@ struct pwr_wkup_data {
 	struct itr_handler *hdl[PWR_NB_WAKEUPPINS];
 	struct itr_handler *parent_hdl;
 	struct stm32_exti_pdata *exti;
-	bool threaded[PWR_NB_WAKEUPPINS];
-	bool pending[PWR_NB_WAKEUPPINS];
 	unsigned int spinlock;
 };
 
@@ -104,48 +102,9 @@ static enum itr_return pwr_it_call_handler(struct pwr_wkup_data *priv,
 	return ITRR_HANDLED;
 }
 
-static enum itr_return pwr_it_threaded_handler(void)
-{
-	struct pwr_wkup_data *priv = pwr_wkup_d;
-	uint32_t i = 0;
-
-	VERBOSE_PWR("threaded_handler");
-
-	for (i = 0; i < PWR_NB_WAKEUPPINS; i++) {
-		if (priv->pending[i]) {
-			VERBOSE_PWR("handle pending wkup irq:%"PRIu32, i);
-			priv->pending[i] = false;
-			pwr_it_call_handler(priv, i);
-		}
-	}
-
-	return ITRR_HANDLED;
-}
-
-static void yielding_stm32_pwr_notif(struct notif_driver *ndrv __unused,
-				     enum notif_event ev)
-{
-	switch (ev) {
-	case NOTIF_EVENT_DO_BOTTOM_HALF:
-		VERBOSE_PWR("PWR Bottom half");
-		pwr_it_threaded_handler();
-		break;
-	case NOTIF_EVENT_STOPPED:
-		VERBOSE_PWR("PWR notif stopped");
-		break;
-	default:
-		EMSG("Unknown event %d", ev);
-		panic();
-	}
-}
-
-struct notif_driver stm32_pwr_notif = {
-	.yielding_cb = yielding_stm32_pwr_notif,
-};
-
 static enum itr_return pwr_it_handler(struct itr_handler *handler)
 {
-	struct pwr_wkup_data *priv = (struct pwr_wkup_data *)handler->data;
+	struct pwr_wkup_data *priv = handler->data;
 	uint32_t i = 0;
 
 	VERBOSE_PWR("it_handler");
@@ -162,12 +121,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 			/* Ack IRQ */
 			io_setbits32(pwr_wkupcr_base(i), WKUPC);
 
-			if (priv->threaded[i] && notif_async_is_started()) {
-				priv->pending[i] = true;
-				notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
-			} else {
-				pwr_it_call_handler(priv, i);
-			}
+			pwr_it_call_handler(priv, i);
 		}
 	}
 
@@ -175,6 +129,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 
 	return ITRR_HANDLED;
 }
+DECLARE_KEEP_PAGER(pwr_it_handler);
 
 static TEE_Result
 stm32_pwr_irq_set_pull_config(size_t it, enum wkup_pull_setting config)
@@ -524,9 +479,6 @@ static TEE_Result stm32mp25_pwr_itr_add(const void *fdt, int wp_node,
 	if (!itr_free)
 		return TEE_ERROR_GENERIC;
 
-	if (hdl->flags & PWR_WKUP_FLAG_THREADED)
-		priv->threaded[it] = true;
-
 	res = gpio_dt_get_by_index(fdt, wp_node, it, "wakeup", &gpio);
 	if (res) {
 		priv->hdl[it] = NULL;
@@ -624,9 +576,6 @@ TEE_Result stm32mp25_pwr_irq_probe(const void *fdt, int node)
 
 	interrupt_enable(priv->parent_hdl->chip, priv->parent_hdl->it);
 
-	if (IS_ENABLED(CFG_CORE_ASYNC_NOTIF))
-		notif_register_driver(&stm32_pwr_notif);
-
 	res = dt_driver_register_provider(fdt, node, NULL,
 					  NULL, DT_DRIVER_NOTYPE);
 	if (res)
diff --git a/core/drivers/stm32mp_pmic2.c b/core/drivers/stm32mp_pmic2.c
index 73143afba..9f5db47e9 100644
--- a/core/drivers/stm32mp_pmic2.c
+++ b/core/drivers/stm32mp_pmic2.c
@@ -682,10 +682,101 @@ static TEE_Result parse_regulator_fdt_nodes(const void *fdt, int node,
 }
 
 #ifdef CFG_STM32_PWR_IRQ
+
+/* We expect a single instance of STPMIC2, if any */
+static struct stpmic2 *stpmic2;
+
+/* When async notif is enabled, bottom half thread must enable the interrupt */
+static bool start_stpmic2_interrupt;
+
+static void pmic_thread_irq_handler(void)
+{
+	assert(stpmic2);
+
+	if (stpmic2_handle_irq(stpmic2))
+		panic("STPMIC2 interrupts");
+}
+
+static void start_stop_stpmic2_interrupt(bool start_not_stop)
+{
+	struct pmic_it_handle_s *prv = NULL;
+
+	assert(stpmic2);
+
+	SLIST_FOREACH(prv, &stpmic2->it_list, link) {
+		if (stpmic2_set_irq_mask(stpmic2, prv->pmic_it,
+					 !start_not_stop)) {
+			EMSG("Failed to %sable STPMIC2 interrupt %u",
+			     start_not_stop ? "en" : "dis", prv->pmic_it);
+			panic();
+		}
+	}
+
+	/* Unmask all over-current interrupts */
+	if (stpmic2_register_write(stpmic2, INT_MASK_R3, 0x00))
+		panic();
+
+	if (stpmic2_register_write(stpmic2, INT_MASK_R4, 0x00))
+		panic();
+}
+
+static void yielding_stm32mp2_stpmic2_notif(struct notif_driver *ndrv __unused,
+					    enum notif_event ev)
+{
+	switch (ev) {
+	case NOTIF_EVENT_DO_BOTTOM_HALF:
+		if (start_stpmic2_interrupt) {
+			IMSG("STPMIC2 notif started");
+			start_stpmic2_interrupt = false;
+			start_stop_stpmic2_interrupt(true /*start_not_stop*/);
+		}
+
+		pmic_thread_irq_handler();
+		break;
+	case NOTIF_EVENT_STOPPED:
+		IMSG("STPMIC2 notif stopped");
+		start_stpmic2_interrupt = false;
+		start_stop_stpmic2_interrupt(false /*!start_not_stop*/);
+		break;
+	default:
+		EMSG("Unknown event %d", ev);
+		panic();
+	}
+}
+
+static bool atomic_stm32mp2_stpmic2_notif(struct notif_driver *ndrv __unused,
+					  enum notif_event ev __maybe_unused)
+{
+	assert(ev == NOTIF_EVENT_STARTED);
+
+	/* Ask for the bottom half to enable PMIC interrupts */
+	start_stpmic2_interrupt = true;
+
+	return true;
+}
+DECLARE_KEEP_PAGER(atomic_stm32mp2_stpmic2_notif);
+
+/* Structure below is not const to prevent it spreads in the unpaged sections */
+static struct notif_driver stm32mp2_stpmic2_notif = {
+	.atomic_cb = atomic_stm32mp2_stpmic2_notif,
+	.yielding_cb = yielding_stm32mp2_stpmic2_notif,
+};
+
+static enum itr_return pmic_it_handler(struct itr_handler *handler __unused)
+{
+	if (notif_async_is_started())
+		notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
+
+	return ITRR_HANDLED;
+}
+DECLARE_KEEP_PAGER(pmic_it_handler);
+
 enum itr_return stpmic2_irq_callback(struct stpmic2 *pmic, uint8_t it_id)
 {
 	struct pmic_it_handle_s *prv = NULL;
 
+	assert(pmic == stpmic2);
+
 	FMSG("Stpmic2 it id %d", (int)it_id);
 
 	SLIST_FOREACH(prv, &pmic->it_list, link)
@@ -726,6 +817,9 @@ static TEE_Result initialize_pmic2_irq(const void *fdt, int node,
 
 	SLIST_INIT(&pmic->it_list);
 
+	if (IS_ENABLED(CFG_CORE_ASYNC_NOTIF))
+		notif_register_driver(&stm32mp2_stpmic2_notif);
+
 	cuint = fdt_getprop(fdt, node, "wakeup-parent", &len);
 	if (!cuint || len != sizeof(uint32_t))
 		panic("Missing wakeup-parent");
@@ -744,8 +838,7 @@ static TEE_Result initialize_pmic2_irq(const void *fdt, int node,
 #if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
 		res = stm32mp25_pwr_itr_alloc_add(fdt, wakeup_parent_node, it,
 						  stpmic2_irq_handler,
-						  PWR_WKUP_FLAG_FALLING |
-						  PWR_WKUP_FLAG_THREADED,
+						  PWR_WKUP_FLAG_FALLING,
 						  pmic, &hdl);
 		if (res)
 			return res;
@@ -754,8 +847,7 @@ static TEE_Result initialize_pmic2_irq(const void *fdt, int node,
 #else
 		res = stm32mp1_pwr_itr_alloc_add(fdt, wakeup_parent_node, it,
 						 stpmic2_irq_handler,
-						 PWR_WKUP_FLAG_FALLING |
-						 PWR_WKUP_FLAG_THREADED,
+						 PWR_WKUP_FLAG_FALLING,
 						 pmic, &hdl);
 		if (res)
 			return res;
@@ -902,6 +994,8 @@ static TEE_Result stm32_pmic2_probe(const void *fdt, int node,
 		free(pmic);
 		return res;
 	}
+
+	stpmic2 = pmic;
 #endif
 
 	res = parse_regulator_fdt_nodes(fdt, node, pmic);
-- 
2.47.2

